Gyroscope

elizaBarnett opened the pull request.

elizaBarnett committed:
added the gyroscope library

elizaBarnett committed:
Basic Gyro Code for turning (needs to add dpad to controls before testing)

elizaBarnett committed:
working on getting the dpad to work

elizaBarnett committed:
Update Gyroscope.java

elizaBarnett committed:
gyroscope

elizaBarnett committed:
changed boolean drivingStraight

elizaBarnett committed:
gyroscope

elizaBarnett committed:
gyroscope code

elizaBarnett committed:
navx

elizaBarnett committed:
created turn command, working on implementing it into the rest of the code

elizaBarnett committed:
gyro code

elizaBarnett committed:
changes  to make oscilate

elizaBarnett committed:
its close to 90 degrees

elizaBarnett committed:
current working gyro code

elizaBarnett committed:
made new control buttons for the turn 180 functions

nbgray commented on src/main/java/frc/robot/commands/GyroTurnCommand.java:
```diff
@@ -0,0 +1,86 @@
+    @Override
+    public void execute() {
+                double error = newAngle - gyro.getGyroAngle();
+                double speed = -error*(0.0027) *.65;
```
All these doubles should be assigned at the top of the class with a comment describing them

elizaBarnett replied:
moved and commented

nbgray commented on src/main/java/frc/robot/commands/GyroTurnCommand.java:
```diff
@@ -0,0 +1,86 @@
+    public void execute() {
+                double error = newAngle - gyro.getGyroAngle();
+                double speed = -error*(0.0027) *.65;
+                if(Math.abs(speed) < .1){
```
Make a constant

elizaBarnett replied:
Done

nbgray commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public void driveStraightGyro(double forwardSpeed,double desiredAngle){ 
+        double error = (desiredAngle-getGyroAngle()) * .0014; 
```
magic number

elizaBarnett replied:
fixed

nbgray commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,69 @@
+
+    public void backToStartingAngle(double speed){
+        double currentAngle = getGyroAngle();
+        double turnAngle = currentAngle % 360;
```
doubles should be x.x to stop an implicit casting, ie 360.0

elizaBarnett replied:
Deleted back to starting angle method in gyro class, made it its own command, mapped it to the b button, and fixed the double formating

elizaBarnett replied:
in the two gyro commands

nbgray commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,69 @@
+    public void backToStartingAngle(double speed){
+        double currentAngle = getGyroAngle();
+        double turnAngle = currentAngle % 360;
+        if( turnAngle/ Math.abs(turnAngle) == 1){
```
it would be better to do 

if( turnAngle < 0.0)
{
speed = -1.0 * speed;
}

If your speed is ever 0 this will have a divide by 0 error.

elizaBarnett replied:
I got ride of this in gyroscope, but I'm working on finishing this issue in the gyro commands

nbgray commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,69 @@
+        }
+        drive.setBrakeMode();
+        while(Math.abs(getGyroAngle()) < Math.abs(currentAngle)){
+            drive.arcadeDrive(-0, -speed); //b/c joysticks come out negative, we can either change the arcade method, or just make all things plugged in negative
```
-0.0 instead of -0

elizaBarnett replied:
done

nbgray commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -9,20 +9,24 @@
   private Controls controls;
+  private Gyroscope gyro;
   private AutonomousCommand autonomousCommand;
-  private TeleopCommand teleopCommand;
```
If you remove teleopCommand how will the robot drive?

nbgray commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -9,20 +9,24 @@
 import frc.robot.subsystems.drive.Drive;
+import frc.robot.subsystems.drive.Gyroscope;
 import frc.robot.commands.AutonomousCommand;
 import frc.robot.commands.TeleopCommand;
```
If you're not using a class, you should remove the import

nbgray commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -101,10 +113,8 @@ public void autonomousPeriodic() {
-      teleopCommand.schedule();
-    }
+    gyro.resetGyro();
+   // drive.setDefaultCommand(new TeleopCommand(controls,drive, gyro));
```
How does the robot drive if there's no teleop command?

elizaBarnett replied:
it glitches occassionally and spins and eventually straightens out again, the encoder drive straight is more reliable. Which means when we merge it I'll save that old code

kushagharahi replied:
I don't think this is how merging will work. If you want the old encoder code you will need to make sure you're not touching that code. Since this branch is ahead of master, the changes in here will take president

elizaBarnett committed:
okay, made a lot of changes, I know the previous pull code was trash, this is better

elizaBarnett committed:
fixed errors

elizaBarnett committed:
flipped sign of speed, works perfectly fine now, gonna try and figure out why

elizaBarnett committed:
changed to full speed version, still worked well

elizaBarnett committed:
got rid of prints and left over useless code

nbgray commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+    // Called repeatedly when this Command is scheduled to run
+    @Override
+    public void execute() {
+                error = gyro.getGyroAngle() % 360.0;
```
Your code is good, but I think the gyro has way too much drift for this to actually work.  But I would be happy to be wrong!

nbgray commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,54 @@
+        startTime = Timer.getFPGATimestamp();        
+        double startAngle = imu.getAngle();
+        try{
+            Thread.sleep(16000);
```
Make this configurable.  Maybe part of the constructor?

elizaBarnett replied:
do you want me to make the sleep configurable?

elizaBarnett committed:
changed sleep to 8 seconds

elizaBarnett committed:
got rid of extra spaces I missed

elizaBarnett committed:
got rid of useless constant (it only slows the robot down if we use it)

nbgray commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -3,20 +3,23 @@
-  private final int LEFT_BACK_CAN_ID = 6;
-  private final int LEFT_FRONT_CAN_ID = 12;
+  private final int LEFT_BACK_CAN_ID = 6;//6
+  private final int LEFT_FRONT_CAN_ID = 5;//12
```
Left is now:

  private final int LEFT_BACK_CAN_ID = 11;
  private final int LEFT_FRONT_CAN_ID = 12;
  //  private final int LEFT_MIDDLE_CAN_ID = 13;

nbgray commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -3,20 +3,23 @@
   // right gear box CAN ids
-  private final int RIGHT_BACK_CAN_ID = 4;
-  private final int RIGHT_FRONT_CAN_ID = 5;
+  private final int RIGHT_BACK_CAN_ID = 8;//4
```
Right is now:

  private final int RIGHT_BACK_CAN_ID = 4;
  private final int RIGHT_FRONT_CAN_ID = 5;
  //  private final int RIGHT_MIDDLE_CAN_ID = 6;

kushagharahi commented:
Can you resolve the conflicts?

kushagharahi commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -31,6 +35,8 @@
   @Log
   private double distance;
+
+  boolean drivingStraight = false;
```
is this used anywhere?

kushagharahi commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -42,9 +48,10 @@ public void robotInit() {
     this.drive = new Drive();
     this.controls = new Controls(new Joystick(JOYSTICK_PORT));
+    this.gyro = new Gyroscope(drive);
+    this.io = new IO(controls,drive, gyro);
```
```suggestion
    this.io = new IO(controls, drive, gyro);
```

kushagharahi commented on src/main/java/frc/robot/commands/TeleopCommand.java:
```diff
@@ -3,38 +3,52 @@
+    boolean drivingStraight = false;
   
-    public TeleopCommand(Controls controls, Drive drive) {
+    public TeleopCommand(Controls controls, Drive drive, Gyroscope gyroscope) {
```
If this drive straight code doesn't work I suggest you revert these changes so you don't overwrite the changes in master.

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,53 @@
+    private double startTime;
+    private double driftPerSecond;
+    //constants
+    private double pValue = .0014;
```
```suggestion
    private final double P_VALUE = .0014;
```

kushagharahi commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -91,34 +100,21 @@ public void autonomousInit() {
   @Override
   public void teleopInit() {
-    this.teleopCommand = new TeleopCommand(controls, drive);
-    if (teleopCommand != null) {
-      teleopCommand.schedule();
-    }
```
This comment: https://github.com/PearceRobotics/infinite-recharge-2020/pull/54/files#r379175534 applies here as well.

kushagharahi commented on src/main/java/frc/robot/commands/GyroTurnCommand.java:
```diff
@@ -0,0 +1,73 @@
+    // Make this return true when this Command no longer needs to run execute()
+    @Override
+    public boolean isFinished() {
+        if(Math.abs(gyro.getGyroAngle() - newAngle) > 1.0)
```
I suggest making this 1.0 deadband a final variable at the top of this class.

kushagharahi commented on src/main/java/frc/robot/commands/TeleopCommand.java:
```diff
@@ -3,38 +3,52 @@
     @Override
     public void execute() {
+        System.out.println(gyroscope.getGyroAngle());
         if (controls.getRightX(DEADZONE) == 0.0) {
```
Having if statements to see what control button you're pressing defeats the purpose of command based programming. In command based programming we want to use the `whenPressed()` function in your IO class in order to let the system schedule the commands. Sorry that we didn't catch this when you merged your encoder code.

The appropriate way to handle this is check your turn modifier value in your arcade drive method and apply your drive straight code if it's equal to 0. It's the same idea except you're not checking your controls. This command should not be aware of the Controls class.

kushagharahi replied:
This might be the reason your drive straight code doesn't work and makes it spin.

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Gyroscope.java:
```diff
@@ -0,0 +1,53 @@
+        startTime = Timer.getFPGATimestamp();
+    }
+
+    public void driveStraightGyro(double forwardSpeed,double desiredAngle){ 
```
instead of your gyro class being aware of the drive class, can the drive class utilize the gyro class instead?

elizaBarnett committed:
fixed can iid

elizaBarnett committed:
fixed things

elizaBarnett committed:
gyro drive straight fixed

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -84,4 +119,8 @@ public double straightTurnPower(){
     return turnPower;
   }
+
+  public double driveStraightGyro(double desiredAngle){  
```
Let’s give method a better name. Maybe something like getAngularDelta or getAngularDrift?

kushagharahi replied:
Give this a better name. Maybe something like getAngularDriftError or getAngularDeltaError?

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -84,4 +119,8 @@ public double straightTurnPower(){
   }
+
+  public double driveStraightGyro(double desiredAngle){  
+    return  (desiredAngle - gyroscope.getGyroAngle()) * P_VALUE;
```
Let’s just move this line to the drive straight part of arcade drive instead of having it in its own method

elizaBarnett committed:
last edits from gyro at meeting

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -3,24 +3,38 @@
+
+  //variables
+  private double desiredAngle;
+  private double turnModifer;
```
You don’t need this variable

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -60,13 +83,25 @@ public void resetEncoders() {
 
   public void setCoastMode(){
+    if(brakeMode == true){
     leftGearbox.setCoastMode();
```
This if statements inner brackets need some tab love

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -60,13 +83,25 @@ public void resetEncoders() {
 
   public void setBrakeMode(){
+    if(brakeMode == false){
     leftGearbox.setBrakeMode();
```
This if statement also needs tab love

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Gearbox.java:
```diff
@@ -1,11 +1,11 @@
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
 
-public class Gearbox {
+public class Gearbox extends SubsystemBase{
```
Since we don’t use the gearboxes directly in a command, does it need to extend SubsystemBase?

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -60,13 +83,25 @@ public void resetEncoders() {
     rightGearbox.setBrakeMode();
+    brakeMode = true;
+    }
+    else{
```
This empty else is not needed

kushagharahi commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -77,6 +84,7 @@ public void robotPeriodic() {
   @Override
   public void autonomousInit() {
     this.autonomousCommand = new AutonomousCommand(distance, maxSpeed, drive);
+    gyro.resetGyro();
```
This isnt needed

kushagharahi commented on src/main/java/frc/robot/commands/GyroTurnCommand.java:
```diff
@@ -0,0 +1,67 @@
+        error = newAngle - gyro.getGyroAngle();
+        speed = -error * P_VALUE;
+
+        /*
```
You were right. Let’s keep this if statement.

kushagharahi commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+                    else if(speed > 0.0) {
+                        speed = minSpeed;
+                    }
+                    else{
```
This empty else isn’t needed

kushagharahi commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public boolean isFinished() {
+        if((Math.abs(gyro.getGyroAngle())% 360.0) > 1.0)
+        {
```
Let’s stay with one bracket style

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -60,13 +83,25 @@ public void resetEncoders() {
   }
 
   public void setCoastMode(){
+    if(brakeMode == true){
```
You don’t actually need to know what mode the gearboxes are. You can just blindly set them to coast and brake mode

kushagharahi commented on src/main/java/frc/robot/subsystems/drive/Drive.java:
```diff
@@ -60,13 +83,25 @@ public void resetEncoders() {
   }
 
   public void setBrakeMode(){
+    if(brakeMode == false){
```
You don’t actually need to know what mode the gearboxes are. You can just blindly set them to coast and brake mode

kushagharahi commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+    private double error; //error the robot is off its original position (from 0 to 360)
+    private double speed;//speed the robot turns
+    //Constants
+    private double ratio = 0.0027;//ratio to multiply error by to get a number between -1 and 1 for the speed
```
Let’s make the same p value changes we made in the gyroturncommand

kushagharahi commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+    private Gyroscope gyro;
+    //Variables
+    private double error; //error the robot is off its original position (from 0 to 360)
+    private double speed;//speed the robot turns
```
This doesn’t need to be a class variabel

kushagharahi commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+    private Drive drive;
+    private Gyroscope gyro;
+    //Variables
+    private double error; //error the robot is off its original position (from 0 to 360)
```
This doesn’t need to be a class variable

kushagharahi commented on src/main/java/frc/robot/commands/ReorientToFieldCommand.java:
```diff
@@ -0,0 +1,67 @@
+    // Make this return true when this Command no longer needs to run execute()
+    @Override
+    public boolean isFinished() {
+        if((Math.abs(gyro.getGyroAngle())% 360.0) > 1.0)
```
Let’s make this 1.0 deadband a final class variable

kushagharahi commented on src/main/java/frc/robot/commands/GyroTurnCommand.java:
```diff
@@ -0,0 +1,67 @@
+    private final double turnAngle;// Angle we want the robot to rotate in degrees
+    private double newAngle;// Angle gyroscope will see new angle as
+    private double error; // continuous error the robot is off from its desired position
+    private double speed;// speed the robot turns
```
These 4 class variables arent needed

kushagharahi commented on src/main/java/frc/robot/commands/GyroTurnCommand.java:
```diff
@@ -0,0 +1,67 @@
+    private double newAngle;// Angle gyroscope will see new angle as
+    private double error; // continuous error the robot is off from its desired position
+    private double speed;// speed the robot turns
+    // Constants
```
Speed and error aren’t needed as class variables.

elizaBarnett committed:
readded min speed, all that needs to be tuned is minspeed, and pvalue

Be hesitant in tuning the deadband. The the larger the deadband the less accurate the turn will be.

elizaBarnett committed:
changed two buttons to work better, started using gyro buttons

elizaBarnett committed:
did almost all of Kusha's comments

Commit:
Merge branch 'gyroscope' of github.com:PearceRobotics/infinite-recharge-2020 into gyroscope

Commit:
Merge branch 'master' into gyroscope

Commit:
fix commands

Commit:
cleanup

Commit:
more cleanup

Commit:
log err

Commit:
more cleanup

Commit:
don't forget to set coast mode
